### PR TITLE
[Backend]: Initial Review 

### PR DESCRIPTION
**Pull request by suraj-aggarwal**
[Backend]: Initial Review

### CODE REVIEW

File: extra/index.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { diamond, equilateral } from './patterns'
+import { users } from './constants'
+import { validation, hasPermission } from './utils'
+
+diamond(process.argv[2])
+equilateral(process.argv[3])
+validation(users)
+let result = hasPermission('getUsers', 'write', 'Trainer')
+console.log(result)
```
Comment (madhavBansal-successiveTech) [inline extra/index.js]: enter required

Comment (madhavBansal-successiveTech) [reply extra/index.js]: console.log("::::::::::::Has Permission:::::::::::",result)

File: extraTs/patterns/equilateral.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            rows = Number(rows);
+            equilateral(rows);
+        }
+    } catch (err) {
+        console.log(err.message);
+    }
+};
+
+
+export default validateEquilateral;
```
Comment (madhavBansal-successiveTech) [inline extraTs/patterns/equilateral.ts]: enter

File: src/controllers/trainee/Controller.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
+import { Request, Response } from 'express';
+import UserRepository from '../../repositories/user/UserRepository';
+import SystemResponse from '../../libs/routes/SystemResponse';
+import { IRequest } from '../../libs/interface';
+import { initSearch } from '../../libs/utils/helper';
+
+class TraineeController {
+  static instance: TraineeController;
+  public userRepo = new UserRepository();
```
Comment (madhavBansal-successiveTech) [inline src/controllers/trainee/Controller.ts:9]: public userRepo:UserRepository = new UserRepository();

File: src/controllers/trainee/Controller.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  private systemResponse: SystemResponse = new SystemResponse();
+  static getInstance(): TraineeController {
+    if (TraineeController.instance instanceof TraineeController) {
```
Comment (madhavBansal-successiveTech) [inline src/controllers/trainee/Controller.ts:12]: if (!TraineeController.instance instanceof TraineeController) {
TraineeController.instance = new TraineeController();
    }
    return TraineeController.instance;

File: src/controllers/trainee/Controller.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+      return TraineeController.instance;
+    }
+    return (TraineeController.instance = new TraineeController());
+  }
+  private constructor() { }
+
+  create = async (req: IRequest, res: Response): Promise<Response> => {
+    try {
+      const userId = req.user.userId;
```
Comment (madhavBansal-successiveTech) [inline src/controllers/trainee/Controller.ts]: const {userId=''} = req.user||{};

File: src/controllers/trainee/Controller.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+      if (result) {
+        delete result._id;
+        return this.systemResponse.success(
+          res,
+          `Trainee added Successfully`,
+          200,
+          result
+        );
+      }
+      return this.systemResponse.success(res, `Unauthorized User`, 403, result);
```
Comment (madhavBansal-successiveTech) [inline src/controllers/trainee/Controller.ts]: How you can say the user is Unauthorized?

File: src/controllers/trainee/Controller.ts [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+  list = async (req: IRequest, res: Response): Promise<Response> => {
+    try {
+      const { skip, limit, sort, search, ...query } = req.query;
+      const options: object = { skip, limit, sort };
+      const filter: object = await initSearch(search);
+      let result;
+      const querySchema = Object.keys(filter).length ? filter : query;
+      result = await this.userRepo.getAllRecord(querySchema, options);
+      if (result.length !== 0) {
```
Comment (madhavBansal-successiveTech) [inline src/controllers/trainee/Controller.ts]: result.length !== 0 can be written as result.length

File: src/controllers/trainee/Controller.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+      });
+    } catch (err) {
+      return this.systemResponse.failure(res, 'No user exits', 200, err);
+    }
+  };
+
+  update = async (req: IRequest, res: Response): Promise<Response> => {
+    try {
+      const { id, dataToUpdate } = req.body;
+      const userId = req.user.userId;
```
Comment (madhavBansal-successiveTech) [inline src/controllers/trainee/Controller.ts]: use destructuring

File: src/controllers/trainee/Controller.ts [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+      return this.systemResponse.failure(res, `can't find record`, 403, result);
+    } catch (err) {
+      return this.systemResponse.failure(res, err.message, 500, err);
+    }
+  };
+
+  delete = async (req: IRequest, res: Response): Promise<Response> => {
+    try {
+      const { id } = req.params;
+      const userId = req.user.userId;
```
Comment (madhavBansal-successiveTech) [inline src/controllers/trainee/Controller.ts]: use destructuring

File: src/controllers/user/Controller.ts [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+  private systemResponse: SystemResponse = new SystemResponse();
+  static getInstance(): UserController {
+    if (UserController.instance instanceof UserController) {
+      return UserController.instance;
+    }
+    return (UserController.instance = new UserController());
+  }
+  private constructor() { }
+
+  private userRepo = new UserRepository();
```
Comment (madhavBansal-successiveTech) [inline src/controllers/user/Controller.ts]: define type of userRepo

File: src/controllers/user/Controller.ts [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+  login = async (req: IRequest, res: Response): Promise<Response> => {
+    try {
+      const email = req.body.email;
+      const password = req.body.password;
```
Comment (madhavBansal-successiveTech) [inline src/controllers/user/Controller.ts]: use destructuring

File: src/controllers/user/Controller.ts [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+      const doc = await this.userRepo.get({ email });
+      if (doc !== null) {
```
Comment (madhavBansal-successiveTech) [inline src/controllers/user/Controller.ts]: doc !== null-->doc

File: src/controllers/user/Controller.ts [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        const match = await bcrypt.compare(password, doc.password);
+        if (!match) {
+          return this.systemResponse.failure(res, 'Invalid Password', 500, {
+            message: 'Enter correct password'
+          });
+        }
+        const id = doc.originalId;
+        const role = doc.role;
```
Comment (madhavBansal-successiveTech) [inline src/controllers/user/Controller.ts]: const id = doc.originalId;
        const role = doc.role;

const {id,role}=doc||{}

File: src/repositories/user/UserRepository.ts [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+  public update = record => {
+    return super.update(record);
+  };
+
+  public isExists = (id: string, email: string) => {
+    const query = { originalId: id, email };
+    return super.isExits(query);
+  };
+
+  public profile = (id: string) => {
```
Comment (madhavBansal-successiveTech) [inline src/repositories/user/UserRepository.ts]: There is no requirement of this function.use get()

File: src/controllers/user/Controller.ts [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+      });
+    }
+  };
+
+  userProfile = async (req: Request, res: Response): Promise<Response> => {
+    try {
+      const token: string = req.headers.authorization;
+      const decodedPayload: any = jwt.verify(token, configuration.SECRET_KEY);
+      const { id } = decodedPayload;
+      const profile = await this.userRepo.profile(id);
```
Comment (madhavBansal-successiveTech) [inline src/controllers/user/Controller.ts]: use get() not profile().Also delete this profile()

File: src/controllers/user/Controller.ts [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+        return this.systemResponse.success(res, `User Profile`, 200, profile);
+      }
+      return this.systemResponse.failure(res, `User Profile not exists`, 403, profile);
+
+    } catch (err) {
+      return this.systemResponse.failure(res, 'failed to fetch profile', 500, err);
+    }
+  };
+
+  isExists = async (id: string, email: string): Promise<boolean> => {
```
Comment (madhavBansal-successiveTech) [inline src/controllers/user/Controller.ts]: No use of this function in Controller file

File: src/libs/Database.ts [hunk 17]
```diff
@@ -0,0 +1,22 @@
+import * as mongoose from 'mongoose';
+import { seedData } from './seedData';
+
+class Database {
+
+    static open = (uri: string) => new Promise((resolve, reject) => {
+        mongoose.connect(uri, { useNewUrlParser: true, useUnifiedTopology: true }, (err) => {
```
Comment (madhavBansal-successiveTech) [inline src/libs/Database.ts:7]: Why using this?

File: src/libs/routes/validationHandler.ts [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+                    if (constant && !req[input][parameter]) {
+                        req[input][parameter] = constant;
+                    }
+                    if (isFieldExits && regex && !regex.test(req[input][parameter]) ) {
+                        errorLogs.push(` ${parameter} is Invalid`);
+                    }
+                });
+            }
+        });
+         (errorLogs.length !== 0) ?  next(errorLogs) : next();
```
Comment (madhavBansal-successiveTech) [inline src/libs/routes/validationHandler.ts]: errorLogs.length !== 0----->errorLogs.length

File: src/repositories/user/UserRepository.ts [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+    return super.get({ id });
+  };
+
+  public get = (query) => {
```
Comment (madhavBansal-successiveTech) [inline src/repositories/user/UserRepository.ts]: use default values for all parameters

File: src/repositories/user/UserRepository.ts [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+    return super.get(query);
+  };
+
+  public getAllRecord = (query: any = {}, options: any = {}) => {
```
Comment (madhavBansal-successiveTech) [inline src/repositories/user/UserRepository.ts]: all functions should be async and define return types of all functions

File: src/repositories/versionable/VersionableRepository.ts [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+    return result;
+  }
+
+  public async get(query: any = {}, options: any = {}): Promise<D> {
+    if (query.id) {
+      query.originalId = query.id;
+      delete query.id;
+    }
+    query = { deletedAt: undefined, ...query };
+    return await this.modelType.findOne(query, options).lean();
```
Comment (madhavBansal-successiveTech) [inline src/repositories/versionable/VersionableRepository.ts]: unnecessary use of await

File: src/repositories/versionable/VersionableRepository.ts [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  public async getAllRecord(query: any = {}, options: any = {}): Promise<D[]> {
+    return await this.modelType.find(
```
Comment (madhavBansal-successiveTech) [inline src/repositories/versionable/VersionableRepository.ts]: unnecessary use of await

File: extra/index.js (hunk 1 above)
Comment (suraj-aggarwal) [reply extra/index.js]: Done

File: src/controllers/trainee/Controller.ts (hunk 3 above)
Comment (suraj-aggarwal) [reply src/controllers/trainee/Controller.ts:9]: done

File: src/controllers/trainee/Controller.ts (hunk 5 above)
Comment (suraj-aggarwal) [reply src/controllers/trainee/Controller.ts]: Done

File: src/controllers/trainee/Controller.ts (hunk 6 above)
Comment (suraj-aggarwal) [reply src/controllers/trainee/Controller.ts]: Done

File: src/controllers/trainee/Controller.ts (hunk 7 above)
Comment (suraj-aggarwal) [reply src/controllers/trainee/Controller.ts]: Done

File: src/controllers/trainee/Controller.ts (hunk 8 above)
Comment (suraj-aggarwal) [reply src/controllers/trainee/Controller.ts]: Done

File: src/controllers/trainee/Controller.ts (hunk 9 above)
Comment (suraj-aggarwal) [reply src/controllers/trainee/Controller.ts]: Done

File: src/controllers/user/Controller.ts (hunk 10 above)
Comment (suraj-aggarwal) [reply src/controllers/user/Controller.ts]: Done

File: src/controllers/user/Controller.ts (hunk 11 above)
Comment (suraj-aggarwal) [reply src/controllers/user/Controller.ts]: Done

File: src/controllers/user/Controller.ts (hunk 12 above)
Comment (suraj-aggarwal) [reply src/controllers/user/Controller.ts]: Done

File: src/controllers/user/Controller.ts (hunk 13 above)
Comment (suraj-aggarwal) [reply src/controllers/user/Controller.ts]: Done

File: src/repositories/user/UserRepository.ts (hunk 14 above)
Comment (suraj-aggarwal) [reply src/repositories/user/UserRepository.ts]: Done

File: src/controllers/user/Controller.ts (hunk 15 above)
Comment (suraj-aggarwal) [reply src/controllers/user/Controller.ts]: Done

File: src/controllers/user/Controller.ts (hunk 16 above)
Comment (suraj-aggarwal) [reply src/controllers/user/Controller.ts]: Done

File: src/libs/routes/validationHandler.ts (hunk 18 above)
Comment (suraj-aggarwal) [reply src/libs/routes/validationHandler.ts]: Done

**Pull request merged**